jbig/pbmtools/jbgtopbm.1
.TH JBGTOPBM 1 "1998-04-10"
.SH NAME
jbgtopbm \- JBIG to portable bitmap file converter
.SH SYNOPSIS
.B jbgtopbm
[
.I options
]
[
.I input-file
| \-  [
.I output-file
]]
.br
.SH DESCRIPTION
Reads in a 
.I JBIG
bi-level image entity (BIE) from a file or standard
input, decompresses it, and outputs a portable bitmap (PBM) file.

.I JBIG
is a highly effective lossless compression algorithm for
bi-level images (one bit per pixel), which is particularly suitable
for scanned document pages.

A
.I JBIG
encoded image can be stored in several resolutions in one or several
BIEs. All resolution layers except the lowest one are stored
efficiently as differences to the next lower resolution layer. Options
.BI -x
and
.BI -y
can be used to stop the decompression at a specified maximal output
image size. The input file can consist of several concatenated BIEs
which contain different increasing resolution layers of the same
image.

If more than one bit per pixel is stored in the JBIG file, then a PGM
file will be produced.
.SH OPTIONS
.TP 14
.BI \-
A single hyphen instead of an input file name will cause 
.I jbgtopbm
to read the data from standard input instead from a file.
.TP
.BI \-x " number"
Decode only up to the largest resolution layer which is still not
more than
.I number
pixels wide. If no such resolution layer exists, then use the smallest
one available.
.TP
.BI \-y " number"
Decode only up to the largest resolution layer which is still not
more than
.I number
pixels high. If no such resolution layer exists, then use the smallest
one available. Options
.BI \-x
and
.BI \-y
can also be used together in which case the largest layer that satisfies
both limits will be selected.
.TP
.BI \-b
Use binary values instead of Gray code words in order to decode pixel
values from multiple bitplanes. This option has only an effect if the
input has more than one bitplane and a PGM output file is produced.
Note that the decoder has to be used in the same mode as the encoder
and cannot determine from the BIE, whether Gray or binary code words
were used by the encoder.
.TP
.BI \-d
Diagnose a BIE. With this option,
.I jbgtopbm
will only print a summary of the header information found in the
input file and then exit.
.TP
.BI \-p " number"
If the input contains multiple bitplanes, then extract only the
specified single plane as a PBM file. The first plane has number 0.
.SH BUGS
Using standard input and standard output for binary data works only on
systems where there is no difference between binary and text streams
(e.g., Unix). On other systems (e.g., MS-DOS), using standard input or
standard output may cause control characters like CR or LF to be
inserted or deleted and this will damage the binary data.
.SH STANDARDS
This program implements the
.I JBIG
image coding algorithm as specified in ISO/IEC 11544:1993 and
ITU-T T.82(1993).
.SH AUTHOR
The
.I jbgtopbm 
program is part of the 
.I JBIG-KIT
package, which has been developed by Markus Kuhn.
The most recent version of this
portable
.I JBIG
library and tools set is freely available on the Internet from
anonymous ftp server ftp.informatik.uni-erlangen.de in directory
pub/doc/ISO/JBIG/. Bug reports should be sent to
<mkuhn@acm.org>.
.SH SEE ALSO
pbm(5), pgm(5), pbmtojbg(1)


NEWS
This file was last updated on April 1, 2009.

Significant changes associated with GraphicsMagick 1.1.15 (April 1, 2009)

  Bugs Fixed:

    o The MagickLibVersion define overflowed as of 1.1.10 since only
      one digit was reserved for the minor version.  Fixed now.

    o MSL: Fix for handling of geometry x,y values in association with
      gravity.

    o TIFF: Ping mode is now fast like it should have been.

    o Conjure subcommand was returning inverted status to the
      environment.

    o Benchmark subcommand was leaking memory for each iteration.

    o Magick++ Image color quantization error measurement was
      performing a wasteful operation.

    o GraphicsMagick.spec was not installing module .la files by default.

    o Bounding box query was failing rather than returning a bounding box
      matching the entire image if there was no subimage to bound.

    o Memory leak/use fixes.

  Security Fixes:

    o Exclude all virtual delegates and coders for pseudo-formats when
      deciding to use file extension as the format.

    o Fix assertion thrown for BMP and DIB files with negative image
      height values.

Significant changes associated with GraphicsMagick 1.1.14 (released June 5, 2008)

  Bugs Fixed:

    o Noise generation was not working correctly for anything but a Q8 build.

    o Poisson noise generator was extremely slow.  Now it is reasonably fast.

    o PDF reader now computes bounding box for rotated PDFs correctly.

  Security Fixes:

    o Cineon reader: Fixed crash with broken file from Sami Liedes.

    o PICT reader: Fixed crash with broken files from Sami Liedes.

    o XCF reader: Fixed crash with broken files from Sami Liedes.

  Security Issue Not Fixed:

    o JP2/JPC reader: Several JPEG-2000 files from Sami Liedes broken collection
      cause crashes in Jasper.  This is not GraphicsMagick's fault.

Significant changes associated with GraphicsMagick 1.1.13 (released May 28, 2008)

  Bugs Fixed:

    o Documentation is now installed according to the conventions established by
      the configure script.  This results in documentation being installed under
      /usr/local/share/doc/GraphicsMagick by default.

    o HWB colorspace now works correctly.

    o Composition with CopyOpacity now produces an image with
      transparency (as expected).

    o Composition now preserves the canvas colorspace.

    o Composition with a displacement map (-displace) no longer leaks an image.

    o Alpha composition now works as expected when both pixels involved
      include transparency.

    o -gamma multiple channel syntax now works as documented.

    o Now compiles with Visual C++ 2008.

Significant changes associated with GraphicsMagick 1.1.12 (released April 28, 2008)

  Security Fixes:

    o Do not access X11 or invoke convenience or stealth delegate
      programs based on the file extension. In particular, these file
      extensions are rejected for consideration as a format specifier:
      'autotrace', 'browse', 'dcraw', 'edit', 'gs-color',
      'gs-color+alpha', 'gs-gray', 'gs-mono', 'launch', 'mpeg-encode',
      'print', 'scan', 'show', 'win', 'xc', and 'x'.

  Bugs Fixed:

    o magick/effect.c: Should now compile for ARM CPU.

    o TIFF: Don't request Kodak private tags since these cause some
      versions of libtiff to misbehave.

    o When performing string expansion of image attribute identifiers, skip
      those which require access to image pixels if image pixels are not
      present.

    o CropImageToHBITMAP(), ImageToHBITMAP(): Fix leak of bitmap handle.

Significant changes associated with GraphicsMagick 1.1.11 (released February 23, 2008)

  Bugs Fixed:

    o BMP: Support large files.

    o DIB: Support large files.

    o PNG: Fix depth handling with 16-bit PNG files in the Q8 build.

    o SUN: Properly report image depth.

    o TIFF: Endian option (-endian) now controls TIFF byte endian order
      rather than bit fill order.

    o DCM, DIB, XBM, XCF, XWD: Eliminate integer overflow vulnerability
      (IDefense 09.19.07).

    o HSL colorspace transform: Avoid optimization bug noticed on Opteron
      with GCC.

    o HWB colorspace transform: Avoid optimization bug noticed on Opteron
      with GCC.

    o RGBTransformImage()/TransformRGBImage(): Was using HWB colorspace
      when HSL was requested.

    o Successfully reads files with names like 'file[123]'.

    o 'gm display': No longer rely on isatty() to determine if input is
      from a pipe (use 'gm display -' to display an image read from a pipe).

  Feature Improvements:

    o 'identify +ping' forces the pixels to be read (similar to GM 1.2).

    o 'gm -version' now indicates if build supports "Large Memory" (i.e. 64-bit).

    o TIFF: Use '-define tiff:fill-order={msb2lsb|lsb2msb}' to control
      TIFF bit fill order.


  Performance Improvements:

    o No longer bogs down if a directory contains hundreds of thousands of
      files and the filename looks like a wildcard specification.


Significant changes associated with GraphicsMagick 1.1.10 (released September 19, 2007)

  Bugs Fixed:

    o Image rotate by -90/270 degrees was producing wrong output.

Significant changes associated with GraphicsMagick 1.1.9 (released September 16, 2007)

  Bugs Fixed:

    o In mogrify command, don't remove file name based on random junk in
      memory.
    o Fixed memory leak when reading MPC files.
    o Fixed crash when writing MIFF format and depth is not expected 8/16/32/.
    o Find latest Ghostscript under Windows which identifies itself as
      "GPL Ghostscript".
    o In mogrify command, don't leak memory in the case where the image
      file contains multiple frames.
    o Fixed crash in PNG and JPEG coders when the image to be written is
      part of an image list.
    o PNG reader errors are now properly reported to the user.
    o TIFF output can now be written to a pipe or other non-seekable
      destination.

jbig/pbmtools/jbgtopbm.c
/*
 *  jbgtopbm - JBIG to Portable Bitmap converter
 *
 *  Markus Kuhn -- mkuhn@acm.org
 *
 *  $Id$
 */

#include <stdio.h>
#include <stdlib.h>
#include <ctype.h>
#include "jbig.h"

#define BUFSIZE 8192

const char usage_msg[] = "JBIGtoPBM converter " JBG_VERSION " -- "
"reads a bi-level image entity (BIE) as input file\n\n"
"usage: %s [<options>] [<input-file> | -  [<output-file>]]\n\n"
"options:\n\n"
"  -x number\tif possible decode only up to a resolution layer not\n"
"\t\twider than the given number of pixels\n"
"  -y number\tif possible decode only up to a resolution layer not\n"
"\t\thigher than the given number of pixels\n"
"  -b\t\tuse binary code for multiple bitplanes (default: Gray code)\n"
"  -d\t\tdiagnose BIE, print content of header and exit\n"
"  -p number\tdecode only one single bit plane (0 = first plane)\n\n";

char *progname;                  /* global pointer to argv[0] */


/*
 * Print usage message and abort
 */
static void usage(void)
{
  fprintf(stderr, usage_msg, progname);
  exit(1);
}


/*
 * Call-back routine for merged image output
 */
void write_it(unsigned char *data, size_t len, void *file)
{
  fwrite(data, len, 1, (FILE *) file);
}


/*
 *
 */
void diagnose_bie(FILE *f)
{
  unsigned char bih[20];
  int len;
  unsigned long xd, yd, l0;

  len = fread(bih, 1, 20, f);
  if (len < 20) {
    printf("Input file is %d < 20 bytes long and does therefore not "
	   "contain an intact BIE header!\n", len);
    return;
  }

  printf("Decomposition of BIH:\n\n  DL = %d\n  D  = %d\n  P  = %d\n"
	 "  -  = %d\n  XD = %lu\n  YD = %lu\n  L0 = %lu\n  MX = %d\n"
	 "  MY = %d\n",
	 bih[0], bih[1], bih[2], bih[3],
	 xd = ((unsigned long) bih[ 4] << 24) | ((unsigned long)bih[ 5] << 16)|
	 ((unsigned long) bih[ 6] <<  8) | ((unsigned long) bih[ 7]),
	 yd = ((unsigned long) bih[ 8] << 24) | ((unsigned long)bih[ 9] << 16)|
	 ((unsigned long) bih[10] <<  8) | ((unsigned long) bih[11]),
	 l0 = ((unsigned long) bih[12] << 24) | ((unsigned long)bih[13] << 16)|
	 ((unsigned long) bih[14] <<  8) | ((unsigned long) bih[15]),
	 bih[16], bih[17]);
  printf("  order   = %d %s%s%s%s%s\n", bih[18],
	 bih[18] & JBG_HITOLO ? " HITOLO" : "",
	 bih[18] & JBG_SEQ ? " SEQ" : "",
	 bih[18] & JBG_ILEAVE ? " ILEAVE" : "",
	 bih[18] & JBG_SMID ? " SMID" : "",
	 bih[18] & 0xf0 ? " other" : "");
  printf("  options = %d %s%s%s%s%s%s%s%s\n", bih[19],
	 bih[19] & JBG_LRLTWO ? " LRLTWO" : "",
	 bih[19] & JBG_VLENGTH ? " VLENGTH" : "",
	 bih[19] & JBG_TPDON ? " TPDON" : "",
	 bih[19] & JBG_TPBON ? " TPBON" : "",
	 bih[19] & JBG_DPON ? " DPON" : "",
	 bih[19] & JBG_DPPRIV ? " DPPRIV" : "",
	 bih[19] & JBG_DPLAST ? " DPLAST" : "",
	 bih[19] & 0x80 ? " other" : "");
  printf("\n  %lu stripes, %d layers, %d planes\n\n",
	 ((yd >> bih[1]) +  ((((1UL << bih[1]) - 1) & xd) != 0) + l0 - 1) / l0,
	 bih[1] - bih[0], bih[2]);

  return;
}


int main (int argc, char **argv)
{
  FILE *fin = stdin, *fout = stdout;
  const char *fnin = "<stdin>", *fnout = "<stdout>";
  int i, j, result;
  int all_args = 0, files = 0;
  struct jbg_dec_state s;
  char *buffer;
  unsigned char *p;
  size_t len, cnt;
  unsigned long xmax = 4294967295UL, ymax = 4294967295UL, max;
  int plane = -1, use_graycode = 1, diagnose = 0;

  buffer = malloc(BUFSIZE);
  if (!buffer) {
    printf("Sorry, not enough memory available!\n");
    exit(1);
  }

  /* parse command line arguments */
  progname = argv[0];
  for (i = 1; i < argc; i++) {
    if (!all_args && argv[i][0] == '-')
      if (argv[i][1] == '\0' && files == 0)
        ++files;
      else
        for (j = 1; j > 0 && argv[i][j]; j++)
          switch(tolower(argv[i][j])) {
          case '-' :
            all_args = 1;
            break;
	  case 'b':
	    use_graycode = 0;
            break;
	  case 'd':
	    diagnose = 1;
            break;
          case 'x':
            if (++i >= argc) usage();
            xmax = atol(argv[i]);
            j = -1;
            break;
          case 'y':
            if (++i >= argc) usage();
            ymax = atol(argv[i]);
            j = -1;
            break;
          case 'p':
            if (++i >= argc) usage();
            plane = atoi(argv[i]);
            j = -1;
            break;
          default:
            usage();
          }
    else
      switch (files++) {
      case 0:
        if (argv[i][0] != '-' || argv[i][1] != '\0') {
          fnin = argv[i];
          fin = fopen(fnin, "rb");
          if (!fin) {
            fprintf(stderr, "Can't open input file '%s", fnin);
            perror("'");
            exit(1);
          }
        }
	if (diagnose) {
	  diagnose_bie(fin);
	  exit(0);
	}
        break;
      case 1:
        fnout = argv[i];
        fout = fopen(fnout, "wb");
        if (!fout) {
          fprintf(stderr, "Can't open input file '%s", fnout);
          perror("'");
          exit(1);
        }
        break;
      default:
        usage();
      }
  }

  /* send input file to decoder */
  jbg_dec_init(&s);
  jbg_dec_maxsize(&s, xmax, ymax);
  result = JBG_EAGAIN;
  do {
    len = fread(buffer, 1, BUFSIZE, fin);
    if (!len) break;
    cnt = 0;
    p = (unsigned char *) buffer;
    while (len > 0 && (result == JBG_EAGAIN || result == JBG_EOK)) {
      result = jbg_dec_in(&s, p, len, &cnt);
      p += cnt;
      len -= cnt;
    }
  } while (result == JBG_EAGAIN || result == JBG_EOK);
  if (ferror(fin)) {
    fprintf(stderr, "Problem while reading input file '%s", fnin);
    perror("'");
    if (fout != stdout) {
      fclose(fout);
      remove(fnout);
    }
    exit(1);
  }
  if (result != JBG_EOK && result != JBG_EOK_INTR) {
    fprintf(stderr, "Problem with input file '%s': %s\n",
	    fnin, jbg_strerror(result, JBG_EN));
    if (fout != stdout) {
      fclose(fout);
      remove(fnout);
    }
    exit(1);
  }
  if (plane >= 0 && jbg_dec_getplanes(&s) <= plane) {
    fprintf(stderr, "Image has only %d planes!\n", jbg_dec_getplanes(&s));
    if (fout != stdout) {
      fclose(fout);
      remove(fnout);
    }
    exit(1);
  }

  if (jbg_dec_getplanes(&s) == 1 || plane >= 0) {
    /* write PBM output file */
    fprintf(fout, "P4\n%ld %ld\n", jbg_dec_getwidth(&s),
	    jbg_dec_getheight(&s));
    fwrite(jbg_dec_getimage(&s, plane < 0 ? 0 : plane), 1,
	   jbg_dec_getsize(&s), fout);
  } else {
    /* write PGM output file */
    if ((size_t) jbg_dec_getplanes(&s) > sizeof(unsigned long) * 8) {
      fprintf(stderr, "Image has too many planes (%d)!\n",
	      jbg_dec_getplanes(&s));
      if (fout != stdout) {
	fclose(fout);
	remove(fnout);
      }
      exit(1);
    }
    max = 0;
    for (i = jbg_dec_getplanes(&s); i > 0; i--)
      max = (max << 1) | 1;
    fprintf(fout, "P5\n%ld %ld\n%lu\n", jbg_dec_getwidth(&s),
	    jbg_dec_getheight(&s), max);
    jbg_dec_merge_planes(&s, use_graycode, write_it, fout);
  }

  /* check for file errors and close fout */
  if (ferror(fout) || fclose(fout)) {
    fprintf(stderr, "Problem while writing output file '%s", fnout);
    perror("'");
    exit(1);
  }

  jbg_dec_free(&s);

  return 0;
}
